src/dbnode/storage/series/buffer.go

```diff
@@ -500,18 +475,205 @@ func (b *dbBuffer) FetchBlocksMetadata(
+
+	return nil, false
+}
+
+func (b *dbBuffer) bucketsAtCreate(
+	t time.Time,
+) *dbBufferBuckets {
+	if buckets, exists := b.bucketsAt(t); exists {
+		return buckets
+	}
+
+	return b.newBucketsAt(t)
+}
+
+func (b *dbBuffer) putBucketInCache(newBuckets *dbBufferBuckets) {
+	replaceIdx := bucketsCacheSize - 1
+	for i, buckets := range b.bucketsCache {
+		// Check if we have the same pointer in cache
+		if buckets == newBuckets {
+			replaceIdx = i
+		}
+	}
+
+	for i := replaceIdx; i > 0; i-- {
+		b.bucketsCache[i] = b.bucketsCache[i-1]
+	}
+
+	b.bucketsCache[0] = newBuckets
+}
+
+func (b *dbBuffer) removeBucketsAt(blockStart time.Time) {
+	buckets, exists := b.bucketsAt(blockStart)
+	if !exists {
+		return
+	}
+
+	b.bucketsPool.Put(buckets)
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+}
+
+type dbBufferBuckets struct {
+	buckets           []*dbBufferBucket
 	start             time.Time
-	encoders          []inOrderEncoder
-	bootstrapped      []block.DatabaseBlock
+	opts              Options
 	lastReadUnixNanos int64
-	drained           bool
+	bucketPool        *dbBufferBucketPool
+}
+
+func (b *dbBufferBuckets) resetTo(
+	start time.Time,
+	opts Options,
+	bucketPool *dbBufferBucketPool,
+) {
+	// nil all elements so that they get GC'd
+	for i := range b.buckets {
+		b.buckets[i] = nil
+	}
+	b.buckets = b.buckets[:0]
+	b.start = start
+	b.opts = opts
+	atomic.StoreInt64(&b.lastReadUnixNanos, 0)
+	b.bucketPool = bucketPool
+}
+
+func (b *dbBufferBuckets) streams(ctx context.Context, wType WriteType) []xio.BlockReader {
+	var res []xio.BlockReader
+	for _, bucket := range b.buckets {
+		if bucket.wType == wType {
+			res = append(res, bucket.streams(ctx)...)
+		}
+	}
+
+	return res
+}
+
+func (b *dbBufferBuckets) streamsLen() int {
+	res := 0
+	for _, bucket := range b.buckets {
+		res += bucket.streamsLen()
+	}
+	return res
+}
+
+func (b *dbBufferBuckets) write(
+	timestamp time.Time,
+	value float64,
+	unit xtime.Unit,
+	annotation []byte,
+	wType WriteType,
+) error {
+	return b.writableBucketCreate(wType).write(timestamp, value, unit, annotation)
+}
+
+func (b *dbBufferBuckets) merge(wType WriteType) (mergeResult, error) {
+	var res mergeResult
+	for _, bucket := range b.buckets {
+		// Only makes sense to merge buckets that are writable
+		if bucket.version == writableBucketVer && wType == bucket.wType {
+			mergeRes, err := bucket.merge()
+			if err != nil {
+				return res, nil
+			}
+			res.merges += mergeRes.merges
+		}
+	}
+
+	return res, nil
+}
+
+func (b *dbBufferBuckets) setLastRead(value time.Time) {
+	atomic.StoreInt64(&b.lastReadUnixNanos, value.UnixNano())
+}
+
+func (b *dbBufferBuckets) lastRead() time.Time {
+	return time.Unix(0, atomic.LoadInt64(&b.lastReadUnixNanos))
+}
+
+func (b *dbBufferBuckets) bootstrap(bl block.DatabaseBlock) {
+	// TODO(juchan): what is a "cold" bootstrap?
+	b.writableBucketCreate(WarmWrite).addBlock(bl)
+}
+
+func (b *dbBufferBuckets) writableBucket(wType WriteType) (*dbBufferBucket, bool) {
```

Comment:
A lot of your logic is dependent on this function, but I think it can mess things up in the case of a failed flush. For example:

```
1) Node bootstraps, stores a block on the writeableBucket
2) Node begins flush, marks the writeable bucket as being version 1 now (not writable)
3) Flush fails
4) Flush begins again, but now it will skip the data from the previous bucket (which was persisted, but to an incomplete flush that will be lost)
```

Reply:
This might just be an issue with the flushing logic....could the flush logic just flush all buckets for a given time that it has? Or does that not work